### PR TITLE
lon lat elevation : validation in NewTelescope.jsx and in the handler if the telescope is fixed

### PR DESCRIPTION
**Pull request by Theodlz**
Solves issue mentionned in https://github.com/skyportal/skyportal/issues/2985, where we would encounter an error because the lon, lat, elevation field are required if the telescope has a fixed location, but that wasn't verified in both frontend and backend.

### CODE REVIEW

**Comment by Theodlz** [comment]
@guynir42 tagging you as i saw you mentionned the error caused by this on slack

**Comment by guynir42** [comment]
Thanks @Theodlz , I will look into it tomorrow morning. 
In the meanwhile we are also adding protection against bad data already in the telescope list (https://github.com/skyportal/skyportal/pull/2987)

File: static/js/components/NewTelescope.jsx [hunk 1]
```diff
@@ -29,11 +29,39 @@ const NewTelescope = () => {
   };
 
   function validate(formData, errors) {
+    console.log(formData);
```
Comment (mcoughlin) [inline static/js/components/NewTelescope.jsx:32]: Remove console statement.

File: static/js/components/NewTelescope.jsx [hunk 2]
```diff
@@ -29,11 +29,39 @@ const NewTelescope = () => {
...
     telescopeList?.forEach((telescope) => {
       if (formData.name === telescope.name) {
         errors.name.addError("Telescope name matches another, please change.");
       }
     });
+    if (formData.fixed_location === true) {
+      if (formData.lon === undefined || formData.lat === undefined) {
```
Comment (mcoughlin) [inline static/js/components/NewTelescope.jsx:39]: Maybe split these up?

File: static/js/components/NewTelescope.jsx [hunk 3]
```diff
@@ -29,11 +29,39 @@ const NewTelescope = () => {
...
+        errors.lon.addError(
+          "Longitude must be specified if telescope is fixed."
+        );
+
+        errors.lat.addError(
+          "Latitude must be specified if telescope is fixed."
+        );
+      } else if (
+        formData.lon < -180 ||
```
Comment (mcoughlin) [inline static/js/components/NewTelescope.jsx:48]: Similar split here?

**Review by mcoughlin — APPROVED** [review]
@Theodlz Looks good. For clarity, would be nice to split up the latitude and longitude reporting / checks.

**Comment by Theodlz** [comment]
> @Theodlz Looks good. For clarity, would be nice to split up the latitude and longitude reporting / checks.

Added it right now @mcoughlin . I think its a good idea for the frontend as it happens all errors in a list to then display all of them. But in the handler, as we return just once. I think they should be grouped by type of error (range of value, undefined), as anyway, if the message is clear (specifying the ranges, or that lon, lat or elevation were undefined) it should be enough for the user to understand.

File: skyportal/handlers/api/telescope.py [hunk 4]
```diff
@@ -40,7 +40,31 @@ def post(self):
         data = self.get_json()
 
         schema = Telescope.__schema__()
-
+        # check if the telescope has a fixed location
+        if 'fixed_location' in data:
+            if data['fixed_location']:
+                if 'lat' not in data or 'lon' not in data or 'elevation' not in data:
+                    return self.error(
+                        'Missing latitude, longitude or elevation, required if the telescope is fixed'
```
Comment (stefanv) [inline skyportal/handlers/api/telescope.py:48]:
```suggestion
                        'Missing latitude, longitude, or elevation; required if the telescope is fixed'
```

File: skyportal/handlers/api/telescope.py [hunk 5]
```diff
@@ -40,7 +40,31 @@ def post(self):
...
+                    )
+                elif (
+                    not isinstance(data['lat'], (int, float))
+                    or not isinstance(data['lon'], (int, float))
+                    or not isinstance(data['elevation'], (int, float))
+                ):
+                    return self.error(
+                        'Latitude, longitude and elevation must all be numbers'
```
Comment (stefanv) [inline skyportal/handlers/api/telescope.py:56]:
```suggestion
                        'Latitude, longitude, and elevation must all be numbers'
```

File: skyportal/handlers/api/telescope.py [hunk 6]
```diff
@@ -40,7 +40,31 @@ def post(self):
...
+                    )
+                elif (
+                    data['lat'] < -90
+                    or data['lat'] > 90
+                    or data['lon'] < -180
+                    or data['lon'] > 180
+                    or data['elevation'] < 0
+                ):
+                    return self.error(
+                        'Latitude must be between -90 and 90, longitude between -180 and 180 and elevation must be positive'
```
Comment (stefanv) [inline skyportal/handlers/api/telescope.py:66]:
```suggestion
                        'Latitude must be between -90 and 90, longitude between -180 and 180, and elevation must be positive'
```

**Review by stefanv — APPROVED** [review]

**Review by mcoughlin — APPROVED** [review]
Thanks @Theodlz

**Pull request merged**